---
title: "prune_and_impute"
author: "Max H"
date: "June 14, 2017"
output: html_document
---

```{r}
suppressMessages(library(SNPRelate))
library(plyr)

setwd("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\")

wheat <- snpgdsOpen("Data\\Intermediate\\GDS\\wheat_phys_subset_sample.gds")
genotypes <- read.gdsn(index.gdsn(wheat, "genotype"))
sample.id <- read.gdsn(index.gdsn(wheat, "sample.id"))
snp.id <- read.gdsn(index.gdsn(wheat, "snp.id"))
snp.chr <- read.gdsn(index.gdsn(wheat, "snp.chromosome"))
snp.pos <- read.gdsn(index.gdsn(wheat, "snp.position"))
samp.annot <- read.gdsn(index.gdsn(wheat, "samp.annot"))
snpgdsClose(wheat)

HRS <- which(samp.annot$designation == "HRS")

for(name in names(samp.annot)) {
  samp.annot[[name]] <- factor(samp.annot[[name]][HRS])
}

snpgdsCreateGeno("Data\\Intermediate\\GDS\\HRS_phys.gds",
                 genmat = genotypes[,HRS],
                 sample.id = sample.id[HRS], 
                 snp.id = snp.id,
                 snp.chromosome = snp.chr,
                 snp.position = snp.pos,
                 other.vars = list(samp.annot = samp.annot),
                 snpfirstdim = T)
```
```{r}
setwd("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\")

wheat <- snpgdsOpen("Data\\Intermediate\\GDS\\HRS_phys.gds")
snp.id <- read.gdsn(index.gdsn(wheat, "snp.id"))
genotypes <- read.gdsn(index.gdsn(wheat, "genotype"))
snp.pos <- read.gdsn(index.gdsn(wheat, "snp.position"))
snp.chr <- read.gdsn(index.gdsn(wheat, "snp.chromosome"))
samp.annot <- read.gdsn(index.gdsn(wheat, "samp.annot"))
sample.id <- as.character(read.gdsn(index.gdsn(wheat, "sample.id")))

IBS <- snpgdsIBS(wheat)
snpgdsClose(wheat)

pairs <- which(IBS$ibs >= 0.99, arr.ind = T)

indices <- vector()
for (i in 1:dim(pairs)[1]) {
  if (pairs[i,1] == pairs[i,2]){
    indices <<- c(indices, i)
  }
}
pairs <- pairs[-indices,]
pairs

NILs <- c("AAC Elie", "Katepwa", "PT434", "BW811", "AC Minto 1",
          "BW275 1", "BW427 1", "BW942", "BW948","Carberry 1", 
          "CDC Stanley 1", "RL4452", "PT754", "Unity", "Somerset 1", 
          "Stettler 1", "AC Michael", "Conway", "Neepawa")

sample.indices <- match(NILs, sample.id)

for(name in names(samp.annot)) {
  samp.annot[[name]] <- factor(samp.annot[[name]][-sample.indices])
}

snpgdsCreateGeno("Data\\Intermediate\\GDS\\HRS_phys_subset_sample.gds",
                 genmat = genotypes[,-sample.indices],
                 sample.id = sample.id[-sample.indices], 
                 snp.id = snp.id,
                 snp.chromosome = snp.chr,
                 snp.position = snp.pos,
                 other.vars = list(samp.annot = samp.annot),
                 snpfirstdim = T)

# save(sample.indices, file = "C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\sample_indices_HRS.Rdata")
```


```{r}
setwd("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\")

wheat <- snpgdsOpen("Data\\Intermediate\\GDS\\HRS_phys_subset_sample.gds")
snp.id <- read.gdsn(index.gdsn(wheat, "snp.id"))
genotypes <- read.gdsn(index.gdsn(wheat, "genotype"))
snp.pos <- read.gdsn(index.gdsn(wheat, "snp.position"))
snp.chr <- read.gdsn(index.gdsn(wheat, "snp.chromosome"))
samp.annot <- read.gdsn(index.gdsn(wheat, "samp.annot"))
sample.id <- as.character(read.gdsn(index.gdsn(wheat, "sample.id")))

set.seed(1000)
snpset.ids.list <- snpgdsLDpruning(wheat, autosome.only = F, 
                                   missing.rate = 0.1, ld.threshold = 1, slide.max.bp = 1e9)
snp.indices <- match(unlist(snpset.ids.list), snp.id)
snpgdsClose(wheat)

snpgdsCreateGeno("Data\\Intermediate\\GDS\\HRS_phys_subset_both.gds",
                 genmat = genotypes[snp.indices,],
                 sample.id = sample.id, 
                 snp.id = snp.id[snp.indices],
                 snp.chromosome = snp.chr[snp.indices],
                 snp.position = snp.pos[snp.indices],
                 other.vars = list(samp.annot = samp.annot),
                 snpfirstdim = T)
```

```{r}
# setwd("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\")
# 
# wheat <- snpgdsOpen("Data\\Intermediate\\GDS\\HRS_phys_subset_both.gds")
# genotypes <- read.gdsn(index.gdsn(wheat, "genotype"))
# sample.id <- as.character(read.gdsn(index.gdsn(wheat, "sample.id")))
# desig <- read.gdsn(index.gdsn(wheat, "samp.annot/designation"))
# dim(genotypes)
# genotypes <- replace(genotypes, genotypes == 0, "A")
# genotypes <- replace(genotypes, genotypes == 2, "G")
# genotypes <- replace(genotypes, genotypes == 3, 0)
# genotypes <- t(genotypes)
# diploid <- cbind(genotypes, genotypes)
# diploid <- diploid[,rep(1:dim(diploid)[2]/2, each = 2) + (0:1) * (dim(diploid)[2]/2)]
# diploid <- cbind(desig, cbind(sample.id, diploid))
# 
# write.table(as.data.frame(diploid), "Data\\Formatted\\HRS_net_struct.csv", quote = F, row.names = F, col.names = F, sep = ",")
```
