---
title: "prune_and_impute"
author: "Max H"
date: "June 14, 2017"
output: html_document
---

```{r}
suppressMessages(library(SNPRelate))
suppressMessages(library(randomForest))
library(plyr)

wheat <- snpgdsOpen("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\wheat_phys.gds")
genotypes <- read.gdsn(index.gdsn(wheat, "genotype"))
sample.id <- read.gdsn(index.gdsn(wheat, "sample.id"))
snp.id <- read.gdsn(index.gdsn(wheat, "snp.id"))
snp.chr <- read.gdsn(index.gdsn(wheat, "snp.chromosome"))
snp.pos <- read.gdsn(index.gdsn(wheat, "snp.position"))
samp.annot <- read.gdsn(index.gdsn(wheat, "samp.annot"))
snpgdsClose(wheat)

HRS <- which(samp.annot$designation == "HRS")

for(name in names(samp.annot)) {
  samp.annot[[name]] <- factor(samp.annot[[name]][HRS])
}

snpgdsCreateGeno("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\HRS_phys.gds",
                 genmat = genotypes[,HRS],
                 sample.id = sample.id[HRS], 
                 snp.id = snp.id,
                 snp.chromosome = snp.chr,
                 snp.position = snp.pos,
                 other.vars = list(samp.annot = samp.annot),
                 snpfirstdim = T)
```

```{r}
wheat <- snpgdsOpen("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\HRS_phys.gds")
genotypes <- read.gdsn(index.gdsn(wheat, "genotype"))
sample.id <- read.gdsn(index.gdsn(wheat, "sample.id"))
snp.id <- read.gdsn(index.gdsn(wheat, "snp.id"))
snp.chr <- read.gdsn(index.gdsn(wheat, "snp.chromosome"))
snp.pos <- read.gdsn(index.gdsn(wheat, "snp.position"))
samp.annot <- read.gdsn(index.gdsn(wheat, "samp.annot"))

set.seed(1000)
snpset.ids.list <- snpgdsSelectSNP(wheat, autosome.only = F, maf = 0.05, missing.rate = 0.1)
snp.indices <- match(unlist(snpset.ids.list), snp.id)
snpgdsClose(wheat)

snpgdsCreateGeno("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\HRS_phys_selectSNP.gds",
                 genmat = genotypes[snp.indices,],
                 sample.id = sample.id, 
                 snp.id = snp.id[snp.indices],
                 snp.chromosome = snp.chr[snp.indices],
                 snp.position = snp.pos[snp.indices],
                 other.vars = list(samp.annot = samp.annot),
                 snpfirstdim = T)
```

```{r}
wheat <- snpgdsOpen("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\HRS_phys_selectSNP.gds")

sample.id <- read.gdsn(index.gdsn(wheat, "sample.id"))
snp.id <- read.gdsn(index.gdsn(wheat, "snp.id"))
snp.pos <- read.gdsn(index.gdsn(wheat, "snp.position"))
snp.chr <- read.gdsn(index.gdsn(wheat, "snp.chromosome"))
samp.annot <- read.gdsn(index.gdsn(wheat, "samp.annot"))

genotypes <- t(read.gdsn(index.gdsn(wheat, "genotype")))
genotypes <- replace(genotypes, genotypes == 3, NA)
genotypes <- replace(genotypes, genotypes == 0, 1)
genotypes <- as.data.frame(apply(genotypes, 2, as.factor))

desig <- read.gdsn(index.gdsn(wheat, "samp.annot/designation"))

snpgdsClose(wheat)

genotypes <- cbind(desig, genotypes)

genotypes.imputed <- rfImpute(desig ~ ., genotypes)
# genotypes.rf <- randomForest(desig ~ ., genotypes.imputed)

genotypes.imputed <- apply(genotypes.imputed[,-1], 2, as.character)
genotypes.imputed <- apply(genotypes.imputed, 2, as.integer)
genotypes.imputed <- replace(genotypes.imputed, genotypes.imputed == 1, 0)

#writing the data out
snpgdsCreateGeno("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\HRS_phys_imputed.gds",
                 genmat = genotypes.imputed,
                 sample.id = sample.id, 
                 snp.id = snp.id,
                 snp.chromosome = snp.chr,
                 snp.position = snp.pos,
                 other.vars = list(samp.annot = samp.annot),
                 snpfirstdim = F)
```



```{r}
wheat <- snpgdsOpen("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\HRS_phys_imputed.gds")

IBS <- snpgdsIBS(wheat)

pairs <- which(IBS$ibs >= 0.99, arr.ind = T)

indices <- vector()
for (i in 1:dim(pairs)[1]) {
  if (pairs[i,1] == pairs[i,2]){
    indices <<- c(indices, i)
  }
}
pairs <- pairs[-indices,]
pairs[1:10,]
```

```{r}
NILs <- c("Katepwa", "BW811", "AC Minto 1", "Avocet 1", "BW275 1", 
          "BW427 1", "BW942", "BW948", "Carberry 1", "CDC Stanley 1",
          "RL4452", "Winalta", "PT754", "Unity", "SWS349", "Somerset 1",
          "Stettler 1", "AC Reed 1", "SWS87", "SWS241", "SWS345",
          "SWS363", "AC Michael", "CDC Makwa", "Conway", "SWS390", 
          "SWS408", "SWS410")

sample.id <- read.gdsn(index.gdsn(wheat, "sample.id"))

sample.indices <- match(NILs, sample.id)

save(sample.indices, file = "C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\sample_indices.Rdata")
```

```{r}
set.seed(1000)
snpset.ids.list <- snpgdsLDpruning(wheat, autosome.only = F, maf = 0.05, missing.rate = 0.1, ld.threshold=0.3, slide.max.bp = 100000)
snp.indices <- match(unlist(snpset.ids.list), snp.id)

save(snp.indices, file = "C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\snp_indices.Rdata")
```

```{r}
genotypes <- read.gdsn(index.gdsn(wheat, "genotype"))
snp.pos <- read.gdsn(index.gdsn(wheat, "snp.position"))
snp.chr <- read.gdsn(index.gdsn(wheat, "snp.chromosome"))
samp.annot <- read.gdsn(index.gdsn(wheat, "samp.annot"))
snpgdsClose(wheat)

for(name in names(samp.annot)) {
  samp.annot[[name]] <- factor(samp.annot[[name]][-sample.indices])
}

snpgdsCreateGeno("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\wheat_phys_imputed_subset.gds",
                 genmat = genotypes[-sample.indices, snp.indices],
                 sample.id = sample.id[-sample.indices], 
                 snp.id = snp.id[snp.indices],
                 snp.chromosome = snp.chr[snp.indices],
                 snp.position = snp.pos[snp.indices],
                 other.vars = list(samp.annot = samp.annot),
                 snpfirstdim = F)

snpgdsCreateGeno("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\wheat_phys_imputed_sample_subset.gds",
                 genmat = genotypes[-sample.indices,],
                 sample.id = sample.id[-sample.indices], 
                 snp.id = snp.id,
                 snp.chromosome = snp.chr,
                 snp.position = snp.pos,
                 other.vars = list(samp.annot = samp.annot),
                 snpfirstdim = F)
```

```{r}
wheat <- snpgdsOpen("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\wheat_phys_imputed_sample_subset.gds")
sample.id <- read.gdsn(index.gdsn(wheat, "sample.id"))
snp.id <- read.gdsn(index.gdsn(wheat, "snp.id"))
genotypes <- read.gdsn(index.gdsn(wheat, "genotype"))
snp.pos <- read.gdsn(index.gdsn(wheat, "snp.position"))
snp.chr <- read.gdsn(index.gdsn(wheat, "snp.chromosome"))
samp.annot <- read.gdsn(index.gdsn(wheat, "samp.annot"))
snpgdsClose(wheat)

HRS <- which(samp.annot$designation == "HRS")

for(name in names(samp.annot)) {
  samp.annot[[name]] <- factor(samp.annot[[name]][HRS])
}

snpgdsCreateGeno("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\HRS_phys_imputed_sample_subset.gds",
                 genmat = genotypes[HRS, ],
                 sample.id = sample.id[HRS], 
                 snp.id = snp.id,
                 snp.chromosome = snp.chr,
                 snp.position = snp.pos,
                 other.vars = list(samp.annot = samp.annot),
                 snpfirstdim = F)


wheat <- snpgdsOpen("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\wheat_phys_imputed_subset.gds")
sample.id <- read.gdsn(index.gdsn(wheat, "sample.id"))
snp.id <- read.gdsn(index.gdsn(wheat, "snp.id"))
genotypes <- read.gdsn(index.gdsn(wheat, "genotype"))
snp.pos <- read.gdsn(index.gdsn(wheat, "snp.position"))
snp.chr <- read.gdsn(index.gdsn(wheat, "snp.chromosome"))
samp.annot <- read.gdsn(index.gdsn(wheat, "samp.annot"))
snpgdsClose(wheat)

HRS <- which(samp.annot$designation == "HRS")

for(name in names(samp.annot)) {
  samp.annot[[name]] <- factor(samp.annot[[name]][HRS])
}

snpgdsCreateGeno("C:\\Users\\Max_H\\OneDrive - University of Guelph\\Pedagogy\\PGDA\\Data\\Formatted\\HRS_phys_imputed_subset.gds",
                 genmat = genotypes[HRS, ],
                 sample.id = sample.id[HRS], 
                 snp.id = snp.id,
                 snp.chromosome = snp.chr,
                 snp.position = snp.pos,
                 other.vars = list(samp.annot = samp.annot),
                 snpfirstdim = F)
```